node/actors/bft/src/chonky_bft/proposer.rs

```diff
@@ -0,0 +1,103 @@
+use crate::{metrics, Config, OutputSender};
+use std::sync::Arc;
+use zksync_concurrency::{ctx, error::Wrap as _, sync, time};
+use zksync_consensus_network::io::ConsensusInputMessage;
+use zksync_consensus_roles::validator;
+
+/// Timeout for creating a proposal. If the proposal is not created in this time, the proposer
+/// will quit trying to create a proposal for this view. This can be different from the replica
+/// timeout for the whole view.
+pub(crate) const PROPOSAL_CREATION_TIMEOUT: time::Duration = time::Duration::milliseconds(2000);
+
+/// The proposer loop is responsible for proposing new blocks to the network. It watches for new
+/// justifications from the replica and if it is the leader for the view, it proposes a new block.
+pub(crate) async fn run_proposer(
+    ctx: &ctx::Ctx,
+    cfg: Arc<Config>,
+    outbound_pipe: OutputSender,
+    mut justification_watch: sync::watch::Receiver<Option<validator::ProposalJustification>>,
+) -> ctx::Result<()> {
+    loop {
+        // Wait for a new justification to be available.
+        let Some(justification) = sync::changed(ctx, &mut justification_watch).await?.clone()
+        else {
+            continue;
+        };
+
+        // If we are not the leader for this view, skip it.
+        if cfg.genesis().view_leader(justification.view().number) != cfg.secret_key.public() {
+            continue;
+        }
+
+        // Create a proposal for the given justification, within the timeout.
+        let proposal = match create_proposal(
+            &ctx.with_timeout(PROPOSAL_CREATION_TIMEOUT),
+            cfg.clone(),
+            justification,
+        )
+        .await
+        {
+            Ok(proposal) => proposal,
+            Err(err) => {
+                tracing::error!("failed to create proposal: {}", err);
+                continue;
+            }
```

Comment:
treat timeout separately from internal errors.

Reply:
Although, if we timeout in block proposal, that's likely an unexpected error.

Reply:
I mean that internal error should be critical and should shut down the component.